samples/bluetooth/peripheral_uart/src/uart_async_adapter.c

```diff
@@ -0,0 +1,551 @@
+
+	data->user_callback = callback;
+	data->user_data = user_data;
+
+	return 0;
+}
+
+static inline void notify_rx_buffer(const struct device *dev)
+{
+	struct uart_event event = {UART_RX_RDY};
+	struct uart_async_adapter_data *data = access_dev_data(dev);
+	bool send = false;
+	k_spinlock_key_t key = k_spin_lock(&(data->lock));
+
+	if (data->rx.buf) {
+
+		__ASSERT_NO_MSG(data->rx.curr_buf >= data->rx.buf);
+		__ASSERT_NO_MSG(data->rx.curr_buf >= data->rx.last_notify_buf);
+		if (data->rx.curr_buf > data->rx.last_notify_buf) {
+
+			event.data.rx.buf = data->rx.buf;
+			event.data.rx.len = data->rx.curr_buf - data->rx.last_notify_buf;
+			event.data.rx.offset = data->rx.last_notify_buf - data->rx.buf;
+			data->rx.last_notify_buf = data->rx.curr_buf;
+			send = true;
+		}
+	}
+	k_spin_unlock(&(data->lock), key);
+	if (send) {
+		LOG_DBG("Notification: UART_RX_RDY (0x%x + %d, size: %d)",
+			(unsigned int)event.data.rx.buf, event.data.rx.offset, event.data.rx.len);
+		user_callback(dev, &event);
+	}
+}
+
+static inline void notify_rx_buffer_release(const struct device *dev, uint8_t *buf)
+{
+	struct uart_event event = {
+		.type = UART_RX_BUF_RELEASED,
+		.data.rx_buf.buf = buf
+	};
+	LOG_DBG("Notification: UART_RX_BUF_RELEASED");
+	user_callback(dev, &event);
+}
+
+static inline void notify_rx_buffer_req(const struct device *dev)
+{
+	struct uart_event event = {
+		.type = UART_RX_BUF_REQUEST,
+	};
+	LOG_DBG("Notification: UART_RX_BUF_REQUEST");
+	user_callback(dev, &event);
+}
+
+static inline void switch_rx_buffer(const struct device *dev, bool request_new)
+{
+	struct uart_async_adapter_data *data = access_dev_data(dev);
+	k_spinlock_key_t key = k_spin_lock(&(data->lock));
+	uint8_t *old_buf = old_buf = data->rx.buf;
+
+	data->rx.buf = data->rx.next_buf;
+	data->rx.curr_buf = data->rx.next_buf;
+	data->rx.last_notify_buf = data->rx.next_buf;
+	data->rx.size_left = data->rx.next_buf_len;
+	data->rx.next_buf = NULL;
+	data->rx.next_buf_len = 0;
+	k_spin_unlock(&(data->lock), key);
+
+	if (old_buf) {
+		notify_rx_buffer_release(dev, old_buf);
+	}
+	if (request_new) {
+		notify_rx_buffer_req(dev);
+	}
+}
+
+static int tx(const struct device *dev, const uint8_t *buf, size_t len, int32_t timeout)
+{
+	int ret = 0;
+	struct uart_async_adapter_data *data = access_dev_data(dev);
+	bool tx_send = false;
+
+	LOG_DBG("%s(%s, 0x%x, %u, %d)", __func__, dev->name, (unsigned int)buf, len, timeout);
+
+	k_spinlock_key_t key = k_spin_lock(&(data->lock));
+
+
+	do {
+		if (!len) {
+			LOG_DBG("%s: no data", __func__);
+			break;
+		}
+		if (data->tx.buf) {
+			ret = -EBUSY;
+			LOG_DBG("%s: busy", __func__);
+			break;
+		}
+		data->tx.buf = buf;
+		data->tx.curr_buf = buf;
+		data->tx.size_left = len;
+		tx_send = true;
+		LOG_DBG("%s: sending", __func__);
+
+	} while (0);
+	k_spin_unlock(&(data->lock), key);
+	if (tx_send) {
+		uart_irq_tx_enable(data->target);
+	}
+
+	if (tx_send && timeout != SYS_FOREVER_MS) {
```

Comment:
`SYS_FOREVER_MS` seems to be introduced after time values were pushed into special structure. So it is used whenever one wants to compare against ints (although there are two cases where people compare against K_FOREVER anyway which is strange). It is used properly.

As for the second part I actually meant to to use use `SYS_TIMEOUT_MS` to convert `timeout` (instead of using `K_MSEC` directly). This is to avoid API mixing.

Reply:
Ok. Technically it would change nothing:
```
/** @brief System-wide macro to convert milliseconds to kernel timeouts
 */
#define SYS_TIMEOUT_MS(ms) ((ms) == SYS_FOREVER_MS ? K_FOREVER : K_MSEC(ms))
```
And as we are checking the value not being SYS_FOREVER_MS... But I get the point.